docs/05-actor-model/idea.md

```diff
@@ -0,0 +1,327 @@
+we need a way, to be able to react on actor finishing his job in
+some universal way, to handle this situation properly in any context.
+
+It is worth noting, that the `Reply` can contain some additional data.
+The id assigned previously is the only required information in the
+`Reply` call, but actor can pass some additional data - `events`
+emitted, which are mostly metadata (to be observed by non-blockchain
+applications mostly), and any arbitrary data it want to pass.
+
+## State
+
+Up until this point we were considering actors as entities performing
+some job, like preparing the meal. If we are considering computer
+programs, such job would be to show something on the screen, maybe
+print something. This is not a case with Smart Contracts. The only
+thing which can be affected by the Smart Contract is their internal
+state. So the state is arbitrary data which is kept by the contract.
+Previously in KFC example I mentioned, that Seller is keeping in mind
+how much actions he scheduled to chiefs are not yet finished - this
+number is actually part of Sellers state.
+
+To give more realistic example of contract state, let think about more
+real life Smart Contract than the restaurant. Let imagine we want
+to create our own currency - maybe we want to create some smart
+contracts based market for some MMORPG game. So we need some way,
+to be able to at least transfer currency between players. We can
+do that, by creating the contract we would call `MmoCurrency`, and
+which would support the `Transfer` action to transfer money to another
+player. Then what would be the state of such contract? It would be
+just a table mapping player names to amount of currency they own. The
+contract we just invited actually exists in CosmWasm examples and it
+is called `cw20-base` contract (it is a bit more complicated, but
+it is basically its core idea).
+
+And now there is a question - how is this helpful to transfer currency
+if I cannot check how much of it do I actually own? It is very good
+question, and the answer to that is simple - whole state of every
+contracts in our system is actually public. It is not universal for
+every Actor model, but it is how it works in CosmWasm, and it is
+kind of forced by the nature of blockchain. Everything happening in
+blockchain has to be public, and if some information should be hidden,
+it have to be stored indirectly.
+
+There is one very important thing about state in CosmWasm, and it is
+state being transactional. Any updates to state are not applied immediately,
+but only when the whole action succeed. It is very important, as
+it guarantees that if something goes wrong in contract, it is always
+left in some proper state. Let consider our `MmoCurrency` case. Imagine,
+that in `Transfer` action we first increase receiver currency amount
+(by updating the state), and only then we decrease the sender amount.
+However before decreasing it, we obviously need to check if sender
```

Comment:
we obviously need to check if a sender possesses enough funds to perform transaction.